### ISSUE TITLE
Error when reverse proxying from apache ssl enabled

### ISSUE THREAD

**Issue by Vinnythetrue** [OliveTin/OliveTin#414]
I am running OliveTin Version: 2024.09.16 on an Ubuntu 22.04 server on port 1337.
If reaching directly with http://$host_ip:1337 no problem everything is working as expected with the HelloWorld default example configuration.

Now I want to give access through the internet using my apache webserver running on 443 :

Here is my sub.domain.com vhost apache configuration (sorry code markups are buggy apparently...) :

```
<Proxy "*localhost:1337*">
        Order deny,allow
        Allow from all
        AuthType Basic
        AuthName "Files"
        AuthBasicProvider external
        AuthExternal pwauth
        Require valid-user
</Proxy>

<VirtualHost *:443>
        ServerName      sub.domain.com
        ServerAdmin     admin@domain.com
        SSLProxyEngine On
        SSLEngine On
        SSLOptions +FakeBasicAuth +ExportCertData +StrictRequire
        SSLCertificateFile /full/path/to/fullchain.pem
        SSLCertificateKeyFile /full/path/to/privkey.pem

         # reverse proxy for olive tin
        ProxyPass / http://localhost:1337/ nocanon KeepAlive=On
        ProxyPassReverse / http://localhost:1337/ KeepAlive=On

        RewriteEngine On
        RewriteCond %{HTTP:Connection} Upgrade [NC]
        RewriteCond %{HTTP:Upgrade} websocket [NC]
        RewriteRule (.*) ws://localhost:1337$1 [P,L]

        ProxyRequests off
        #ProxyVia on      

        AllowEncodedSlashes NoDecode
</VirtualHost>
```

I can see the OliveTin page with the hello button through Apache.

The issue begins when clicking the button, therefore calling /api/StartAction on olivetin. Apache log gives me the error

```
[Fri Sep 27 10:32:05.233201 2024] [proxy_http:error] [pid 102160] (20014)Internal error (specific information not available): [client 192.168.0.254:50498] AH01102: error reading status line from remote server localhost:1337, referer: https://test.vadhe.net/
[Fri Sep 27 10:32:05.233297 2024] [proxy_http:debug] [pid 102160] mod_proxy_http.c(1163): [client 192.168.0.254:50498] AH01105: NOT Closing connection to client although reading from backend server localhost:1337 failed., referer: https://test.vadhe.net/
[Fri Sep 27 10:32:05.233327 2024] [proxy:error] [pid 102160] [client 192.168.0.254:50498] AH00898: Error reading from remote server returned by /api/StartAction, referer: https://test.vadhe.net/
[Fri Sep 27 10:32:05.233349 2024] [proxy:debug] [pid 102160] proxy_util.c(2553): AH00943: http: has released connection for (localhost)
```

The curl request on client side looks like this :

```
curl 'https://sub.domain.com/api/StartAction' \
  -H 'Accept: */*' \
  -H 'Accept-Language: fr-FR,fr;q=0.9,en-US;q=0.8,en;q=0.7' \
  -H 'Authorization: Basic XXXXXXXX' \
  -H 'Connection: keep-alive' \
  -H 'Content-Type: application/json' \
  -H 'Origin: https://test.vadhe.net' \
  -H 'Referer: https://test.vadhe.net/' \
  -H 'Sec-Fetch-Dest: empty' \
  -H 'Sec-Fetch-Mode: cors' \
  -H 'Sec-Fetch-Site: same-origin' \
  -H 'User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/123.0.0.0 Safari/537.36' \
  -H 'sec-ch-ua: "Google Chrome";v="123", "Not:A-Brand";v="8", "Chromium";v="123"' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'sec-ch-ua-platform: "Linux"' \
  --data-raw '{"actionId":"e94b2212-f580-4d50-a101-79bbfa029743","arguments":[],"uniqueTrackingId":"01241155-ad8f-4692-8519-62b978b20e69"}'
```

Running this obviously fails with an apache 502 error. If I manipulate the curl to call directly the OliveTin app, I can see it returns an empty response. While the OliveTin log gives me this error :

```
Sep 27 10:35:51 tyler OliveTin[103029]: level="debug" msg="UserFromContext" usergroup="guest" username="guest"
Sep 27 10:35:51 tyler OliveTin[103029]: level="debug" msg="UserFromContext" usergroup="guest" username="guest"
Sep 27 10:35:51 tyler OliveTin[103029]: level="debug" msg="ACL check - isAllowedView" actionTitle="Hello world!" relevantAcls="0" requiredPermission="1" usergroup="guest" username="guest"
Sep 27 10:35:51 tyler OliveTin[103029]: level="debug" msg="ACL check - isAllowedExec" actionTitle="Hello world!" relevantAcls="0" requiredPermission="2" usergroup="guest" username="guest"
Sep 27 10:35:56 tyler OliveTin[103029]: level="debug" msg="Adding CORS header origin: \"https://test.vadhe.net\""
Sep 27 10:35:56 tyler OliveTin[103029]: level="debug" msg="UserFromContext" usergroup="guest" username="guest"
Sep 27 10:35:56 tyler OliveTin[103029]: panic: runtime error: invalid memory address or nil pointer dereference
Sep 27 10:35:56 tyler OliveTin[103029]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x9da438]
Sep 27 10:35:56 tyler OliveTin[103029]: goroutine 82 [running]:
Sep 27 10:35:56 tyler OliveTin[103029]: github.com/OliveTin/OliveTin/internal/grpcapi.(*oliveTinAPI).StartAction(0xc0000a02d8, {0xcbae60, 0xc0003a3710}, 0xc0001d2d80)
Sep 27 10:35:56 tyler OliveTin[103029]:         /home/runner/work/OliveTin/OliveTin/internal/grpcapi/grpcApi.go:73 +0x1d8
Sep 27 10:35:56 tyler OliveTin[103029]: github.com/OliveTin/OliveTin/gen/grpc._OliveTinApiService_StartAction_Handler({0xb74d20?, 0xc0000a02d8}, {0xcbae60, 0xc0003a3710}, 0xc0001cab00, 0x0)
Sep 27 10:35:56 tyler OliveTin[103029]:         /home/runner/work/OliveTin/OliveTin/gen/grpc/OliveTin_grpc.pb.go:295 +0x169
Sep 27 10:35:56 tyler OliveTin[103029]: google.golang.org/grpc.(*Server).processUnaryRPC(0xc0000df400, {0xcbae60, 0xc0003a3680}, {0xcbf068, 0xc0002c6680}, 0xc0003967e0, 0xc0001c1470, 0x1163338, 0x0)
Sep 27 10:35:56 tyler OliveTin[103029]:         /home/runner/go/pkg/mod/google.golang.org/grpc@v1.62.1/server.go:1386 +0xe23
Sep 27 10:35:56 tyler OliveTin[103029]: google.golang.org/grpc.(*Server).handleStream(0xc0000df400, {0xcbf068, 0xc0002c6680}, 0xc0003967e0)
Sep 27 10:35:56 tyler OliveTin[103029]:         /home/runner/go/pkg/mod/google.golang.org/grpc@v1.62.1/server.go:1797 +0x100c
Sep 27 10:35:56 tyler OliveTin[103029]: google.golang.org/grpc.(*Server).serveStreams.func2.1()
Sep 27 10:35:56 tyler OliveTin[103029]:         /home/runner/go/pkg/mod/google.golang.org/grpc@v1.62.1/server.go:1027 +0x8b
Sep 27 10:35:56 tyler OliveTin[103029]: created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 50
Sep 27 10:35:56 tyler OliveTin[103029]:         /home/runner/go/pkg/mod/google.golang.org/grpc@v1.62.1/server.go:1038 +0x135
Sep 27 10:35:56 tyler systemd[1]: OliveTin.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
```

Now if I create a curl request from a direct access (that works), and do the opposite, meaning changing the url to https://sub.domain.com and adding the Authorization header to manually make it pass through apache, guess what ? It works ! Here is an example of such a curl request :

```
curl 'http://$host_ip:1337/api/StartAction' \
  -H 'Accept: */*' \
  -H 'Authorization: Basic XXXXXXX' \
  -H 'Accept-Language: fr-FR,fr;q=0.9,en-US;q=0.8,en;q=0.7' \
  -H 'Connection: keep-alive' \
  -H 'Content-Type: application/json' \
  -H 'Origin: http://$host_ip:1337' \
  -H 'Referer: http://$host_ip:1337/' \
  -H 'User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/123.0.0.0 Safari/537.36' \
  --data-raw '{"actionId":"e769021d-0792-4d4b-bbca-10a9af932e07","arguments":[],"uniqueTrackingId":"1727433469220"}' \
  --insecure
```

At this point I have noticed some differences :
- uniqueTrackingId i differently formed
- actionId is different, is this normal ?

I am quite stuck here as I don't understand what is going on and what is wrong with my setup...

**Comment by jamesread** [comment]
Heya :-) 

So, a couple of things. OliveTin is crashing because of a bug where if it cannot find the actionId from the API, it will generate a Null Pointer, and segfault (SIGSEGV) - you can see this at your logs, it's crashing in grpcApi.go, line 73. I just patched this tonight, because I found it as well when I was setting up my personal OliveTin instance to do some new stuff! Commit 9723a38ca1266b6e81d28e59e2c444651229aab4 contains this fix (I just started a build, 2024.10.01, which should be out in about 30-45 minutes from this post).

However, the reason it's not finding the action, is because it's generating unique UUIDs every time it starts up - we can see this from your "actionId" requests. I imagine you're probably using your browser's web developer console or similar to get those actionIds?

Instead, in your config, manually set a static ID on actions you want to call from the API, that will make your life somewhat easier!;

```yaml
actions:
  - title: Date
    shell: date
    id: date
```

Then your actionId will just be `date`. 

That should make your testing a lot easier, and eliminate some of your problems. 

Regarding your Apache config is generally looks ok, other than you're not handling the websocket connection quite how I do it in the docs. In principle it looks ok, but I won't go ahead and test yet until we've fixed the two above issues (API crash - you should upgrade to 2024.10.01 when it's out), and setting a static `id` on your actions.

**Comment by Vinnythetrue** [comment]
Hey !

Thanks for your time and answer !

In the meantime, I found what problem I was having : caching !

So, what I did :
- still running the same OliveTin version
- didn't add fixed ids
- added strict no-caching configuration in apache :
```
<IfModule mod_headers.c>
	    Header unset ETag
	    Header set Cache-Control "max-age=0, no-cache, no-store, must-revalidate"
	    Header set Pragma "no-cache"
	    Header set Expires "Wed, 12 Jan 1980 05:00:00 GMT"
</IfModule>```
